components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F

```diff
@@ -1671,278 +1730,315 @@ end subroutine ocn_diagnostic_solve_GMvel!}}}
-            do k = minLevelCell(iCell), maxLevelCell(iCell)
-               r_tmp = dvEdge_temp * normalVelocity(k, iEdge) * invAreaCell1
+            do k = kmin,kmax
+               r_tmp = dvEdge_temp*normalVelocity(k,iEdge)*invAreaCell1
 
-               divergence(k, iCell) = divergence(k, iCell) - edgeSignOnCell_temp * r_tmp
-               div_hu(k) = div_hu(k) - layerThicknessEdgeFlux(k, iEdge) * edgeSignOnCell_temp * r_tmp
-               kineticEnergyCell(k, iCell) = kineticEnergyCell(k, iCell) &
-                                              + 0.25 * r_tmp * dcEdge_temp * normalVelocity(k,iEdge)
-               ! Compute vertical velocity from the horizontal total transport
+               divergence(k,iCell) = divergence(k,iCell) &
+                                   - edgeSignOnCell_temp*r_tmp
+               div_hu(k) = div_hu(k) &
+                         - layerThicknessEdgeFlux(k,iEdge)* &
+                           edgeSignOnCell_temp*r_tmp
                div_huTransport(k) = div_huTransport(k) &
-                                    - layerThicknessEdgeFlux(k, iEdge) * edgeSignOnCell_temp &
-                                    * dvEdge_temp * normalTransportVelocity(k, iEdge) * invAreaCell1
+                                  - layerThicknessEdgeFlux(k,iEdge)* &
+                                    edgeSignOnCell_temp*dvEdge_temp* &
+                                    normalTransportVelocity(k,iEdge)* &
+                                    invAreaCell1
+               kineticEnergyCell(k,iCell) = kineticEnergyCell(k,iCell) &
+                                          + 0.25*r_tmp*dcEdge_temp* &
+                                            normalVelocity(k,iEdge)
             end do
          end do
-         ! Vertical velocity at bottom (maxLevelCell(iCell)+1) is zero, initialized above.
-         vertVelocityTop(1:minLevelCell(iCell)-1, iCell) = 0.0_RKIND
-         vertVelocityTop(maxLevelCell(iCell)+1, iCell) = 0.0_RKIND
-         do k = maxLevelCell(iCell), 1, -1
-            vertVelocityTop(k,iCell) = vertVelocityTop(k+1,iCell) - div_hu(k)
-            vertTransportVelocityTop(k,iCell) = vertTransportVelocityTop(k+1,iCell) - div_huTransport(k)
+         ! Vertical velocity at bottom is zero, initialized above.
+         vertVelocityTop(1:kmin-1,iCell) = 0.0_RKIND
+         vertVelocityTop(kmax+1  ,iCell) = 0.0_RKIND
+         vertTransportVelocityTop(1:kmin-1,iCell) = 0.0_RKIND
+         vertTransportVelocityTop(kmax+1  ,iCell) = 0.0_RKIND
+         do k = kmax, 1, -1
+            vertVelocityTop(k,iCell) = &
+            vertVelocityTop(k+1,iCell) - div_hu(k)
+            vertTransportVelocityTop(k,iCell) = &
+            vertTransportVelocityTop(k+1,iCell) - div_huTransport(k)
          end do
       end do
 #ifndef MPAS_OPENACC
       !$omp end do
-      !$omp end parallel
 #endif
 
+      !$acc exit data delete(div_hu, div_huTransport)
       deallocate(div_hu,div_huTransport)
```

Comment:
@philipwjones when I run this with two OpenMP threads, I get the error 
```
217 Error termination. Backtrace:
218 At line 2015 of file mpas_ocn_diagnostics.F
219 Fortran runtime error: Attempt to DEALLOCATE unallocated 'div_hu'
```
but it works fine for single thread cases. Is the allocate/deallocate nesting with OpenMP incorrect?

Reply:
Sorry about that - looks like I was going to merge the two parallel regions, but forgot to deal with the intervening code between loops. Just committed a change to reinsert the parallel directives as before.